Add aot_eager_then_compile stance

bobrenjc93 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #148530
* __->__ #148509

Sometimes `eager_then_compile` stance isn't enough since some models are so close to the memory limit that going to eager will OOM since we don't get the memory reductions from activation checkpointing. This PR introduces `aot_eager_then_compile` which avoids the expensive inductor compile, but still does aot_eager to get the benefits of memory reduction in the first invocation.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bobrenjc93 committed:
Add aot_eager_then_compile stance

[ghstack-poisoned]

bobrenjc93 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -203,7 +203,9 @@ def eager_then_compile(*args, **kwargs):
 
-                compiler_fn = callback._torchdynamo_orig_callable._torchdynamo_orig_callable.compiler_fn
+                compiler_fn = (
+                    callback._torchdynamo_orig_callable._torchdynamo_orig_callable
```
I realized that this is much more robust since sometimes `compiler_fn` doesn't exist, but you always can just invoke the callable directly.

williamwen42 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -216,6 +218,40 @@ def eager_then_compile(*args, **kwargs):
+    elif _stance.stance == "aot_eager_then_compile":
+        if callback not in (False, None):
+
+            def aot_eager_then_compile(*args, **kwargs):
```
Can we factor out common code with `def eager_then_compile`?

bobrenjc93 committed:
Update on "Add aot_eager_then_compile stance"


Sometimes `eager_then_compile` stance isn't enough since some models are so close to the memory limit that going to eager will OOM since we don't get the memory reductions from activation checkpointing. This PR introduces `aot_eager_then_compile` which avoids the expensive inductor compile, but still does aot_eager to get the benefits of memory reduction in the first invocation.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Add aot_eager_then_compile stance"


Sometimes `eager_then_compile` stance isn't enough since some models are so close to the memory limit that going to eager will OOM since we don't get the memory reductions from activation checkpointing. This PR introduces `aot_eager_then_compile` which avoids the expensive inductor compile, but still does aot_eager to get the benefits of memory reduction in the first invocation.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Add aot_eager_then_compile stance"


Sometimes `eager_then_compile` stance isn't enough since some models are so close to the memory limit that going to eager will OOM since we don't get the memory reductions from activation checkpointing. This PR introduces `aot_eager_then_compile` which avoids the expensive inductor compile, but still does aot_eager to get the benefits of memory reduction in the first invocation.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 committed:
Update on "Add aot_eager_then_compile stance"


Sometimes `eager_then_compile` stance isn't enough since some models are so close to the memory limit that going to eager will OOM since we don't get the memory reductions from activation checkpointing. This PR introduces `aot_eager_then_compile` which avoids the expensive inductor compile, but still does aot_eager to get the benefits of memory reduction in the first invocation.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bobrenjc93 commented:
@Pytorchbot merge

bobrenjc93 commented:
@Pytorchbot merge

bobrenjc93 commented:
@pytorchbot merge

bobrenjc93 commented:
@pytorchbot merge
